src/Microsoft.AspNet.OData.Shared/EdmChangedObjectCollection.cs

```diff
@@ -60,6 +70,321 @@ private void Initialize(IEdmEntityType entityType)
             _entityType = entityType;
             _edmType = new EdmDeltaCollectionType(new EdmEntityTypeReference(_entityType, isNullable: true));
             _edmTypeReference = new EdmCollectionTypeReference(_edmType);
+            
+        }
+
+        /// <summary>
+        /// Patch for Types without underlying CLR types
+        /// </summary>
+        /// <param name="originalCollection"></param>
+        /// <returns>ChangedObjectCollection response</returns>
+        internal EdmChangedObjectCollection Patch(ICollection<IEdmStructuredObject> originalCollection)
+        {
+            EdmODataAPIHandler apiHandler = new DefaultEdmODataAPIHandler(originalCollection, _entityType);
+
+            return CopyChangedValues(apiHandler);
+        }
+
+        /// <summary>
+        /// Patch for EdmChangedObjectCollection, a collection for IEdmChangedObject 
+        /// </summary>
+        /// <returns>ChangedObjectCollection response</returns>
+        internal EdmChangedObjectCollection Patch(EdmODataAPIHandler apiHandler, ODataEdmAPIHandlerFactory apiHandlerFactory)
+        {
+            return CopyChangedValues(apiHandler, apiHandlerFactory);
+        }
+
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Performance", "CA1800:DoNotCastUnnecessarily")]
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes")]
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Maintainability", "CA1506:AvoidExcessiveClassCoupling")]
+        internal EdmChangedObjectCollection CopyChangedValues(EdmODataAPIHandler apiHandler, ODataEdmAPIHandlerFactory apiHandlerFactory = null)
+        {
+            EdmChangedObjectCollection changedObjectCollection = new EdmChangedObjectCollection(_entityType);
+            IEdmStructuralProperty[] keys = _entityType.Key().ToArray();
```

Comment:
I think there must be a better approach than this. I was initially suspicious because of the `ToArray` call, but then I saw it makes sense because it is possible for `keys` to be enumerated multiple times. However, at least for the failed case, we are doing the same `.Name` deference for the keys. Maybe here we should do `_entityType.Key().Select(key => key.Name).ToArray()` so that we don't have to deference the name so many times?

Reply:
Good one